### PR TITLE
SqueezeNet Implentation

### PR DESCRIPTION
**Pull request by shubham1206agra**
Implemented SqueezeNet using the new ann-vtable structure as a Layer for more straightforward serialization and use.

### CODE REVIEW

**Review by kartikdutt18 — COMMENTED** [review]
Looks good to me, can we add some test files for this model as well

File: models/squeezenet/squeezenet.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+  //! Virtual destructor: delete all held layers.
+  virtual ~SqueezeNetType()
+  { /* Nothing to do here. */ }
+
+  //! Create a copy of the SqueezeNetType (this is safe for polymorphic use).
+  SqueezeNetType* Clone() const { return new SqueezeNetType(*this); }
+
+  /**
+   * Get the FFN object representing the network.
```
Comment (rcurtin) [inline models/squeezenet/squeezenet.hpp:85]: Don't forget to document here that the caller is responsible for deleting the layer---unless they call `Add()` to add the network directly to an `FFN` (where the FFN will handle deleting the layer).

File: models/squeezenet/squeezenet_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    Fire(32, 128, 128);
+    this->template Add<ann::MaxPooling>(3, 3, 2, 2, false);
+    Fire(48, 192, 192);
+    Fire(48, 192, 192);
+    Fire(64, 256, 256);
+    Fire(64, 256, 256);
+  }
+  else
+  {
+    mlpack::Log::Fatal << "Unsuppoted SqueezeNet version." << std::endl;
```
Comment (rcurtin) [inline models/squeezenet/squeezenet_impl.hpp:153]: `Unsupported`

File: tests/squeezenet_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+template<typename LayerType>
+void ModelSerializationTest(LayerType& layer)
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output(1000, 10, arma::fill::randu);
+
+  ann::FFN<> model;
+  model.Add<LayerType>(layer);
+
+	model.InputDimensions() = std::vector<size_t>({224, 224, 3});
```
Comment (rcurtin) [inline tests/squeezenet_tests.cpp:84]: Here and some other places there are style issues---be sure to use spaces and not tabs. :+1:

**Review by rcurtin — COMMENTED** [review]
Everything looks correct to me here, great work! Just a couple small comments.

**Review by rcurtin — APPROVED** [review]
Nice, thanks for handling the comments. :+1:

**Pull request merged**